### DISCUSSION TITLE
Validate missing or extra properties in json messages with jsonschema

### DISCUSSION THREAD

**Discussion by jon-chuang** [foxglove/community#662] [Q&A]
**Description**
The behaviour upon parsing with JSONSchema should:
1. If fields are missing, either null pad values or throw error
2. If extra fields are present, throw error or ignore

**Steps To Reproduce**
```js
const channel = parseChannel({
  messageEncoding: "json",
  schema: {
    name: "X",
    encoding: "jsonschema",
    data: new TextEncoder().encode(
      JSON.stringify({ type: "object", properties: { value: { type: "integer" } } }),
    ),
  },
});
assert channel.deserializer(new TextEncoder().encode(JSON.stringify({notValue: "not integer" })))) === { notValue: "not integer" }'
```

**Expected Behavior**
Throw useful error report for user action.

Currently, we do none of the above. The required fields are not null padded but `undefined` (although I prefer to throw an error for missing fields, except if they are declared as nullable, in which case we null-pad), while the extra fields are not ignored and do not throw error.

**Comment by foxhubber** [comment]
Internal tracking ticket: [FG-2521](https://linear.app/foxglove/issue/FG-2521/bugmcap-support-decoding-json-message-with-jsonschema-with-schema)

**Comment by jtbandes** [comment]
Converting this to a discussion. This is not a bug per se, because the app does not attempt nor claim to perform any schema validation today (unless it is required in order to parse message data).

I would be concerned about the performance implications of validating every parsed message. One option we've discussed in the past is to maybe validate only the _first_ message on each channel. This would handle the common case where a channel maintains the same shape from message to message, without a performance penalty.
